### PR TITLE
[ORCA-4600] Add support for cache variables

### PR DESCRIPTION
**Pull request by c-kaieong**
Add CRUD support for Event Orchestration Cache Variables

### CODE REVIEW

File: pagerduty/event_orchestration_cache_variable.go
Comment (alenapan) [inline pagerduty/event_orchestration_cache_variable.go]: I think `omitempty` needs to be removed from the `Disabled` property because it drops "falsy" fields from the payload. And here is what's classified as a falsy field, according to the [docs](https://pkg.go.dev/encoding/json#:~:text=The%20%22omitempty%22%20option%20specifies%20that,the%20field%20is%20always%20omitted.):
> The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any empty array, slice, map, or string.

So basically if they want to send `disabled: false` they won't be able to do it. We had a similar issue with EOs where a customer created a Router with disabled rules first, then tried setting them to `disabled: false`, and the new values kept getting ignored because they were never sent to the API.

File: pagerduty/event_orchestration_cache_variable.go
Comment (alenapan) [inline pagerduty/event_orchestration_cache_variable.go]: And same with `Conditions` - because we want to be able to send empty arrays we should remove `omitempty`.

File: pagerduty/event_orchestration_cache_variable_test.go
Comment (alenapan) [inline pagerduty/event_orchestration_cache_variable_test.go]: Typo in `TestGlobalOrchestrationICacheVariableUpdate`
```suggestion
func TestGlobalOrchestrationCacheVariableUpdate(t *testing.T) {
```

File: pagerduty/event_orchestration_cache_variable_test.go
Comment (alenapan) [inline pagerduty/event_orchestration_cache_variable_test.go]: Typo in `TestGlobalOrchestrationIntegrationDelete`
```suggestion
func TestGlobalOrchestrationCacheVariableDelete(t *testing.T) {
```

File: pagerduty/event_orchestration_cache_variable_test.go
Comment (alenapan) [inline pagerduty/event_orchestration_cache_variable_test.go]: **Nit**: Let's maybe change the `oId` here and in all service Cache Var tests below to a valid obfuscated service ID to reflect the real world scenario more accurately?

File: pagerduty/event_orchestration_cache_variable_test.go
Comment (alenapan) [inline pagerduty/event_orchestration_cache_variable_test.go]: Typo in `TestServiceOrchestrationICacheVariableUpdate`
```suggestion
func TestServiceOrchestrationCacheVariableUpdate(t *testing.T) {
```

File: pagerduty/event_orchestration_cache_variable_test.go
Comment (alenapan) [inline pagerduty/event_orchestration_cache_variable_test.go]: Typo in `TestServiceOrchestrationIntegrationDelete`
```suggestion
func TestServiceOrchestrationCacheVariableDelete(t *testing.T) {
```

**Comment by c-kaieong** [comment]
@alenapan Ah, thanks for catching all those subtle typos and the clarification around `omitempty`! I've made the changes in https://github.com/heimweh/go-pagerduty/pull/149/commits/5279d06491253722cfc249e22c52166ea4e2c104

**Review by alenapan — APPROVED** [review]
LGTM 👍 Probably worth it to leave it as a draft PR since we can't merge this until the Public API docs go live

File: pagerduty/event_orchestration_cache_variable.go
Comment (imjaroiswebdev) [inline pagerduty/event_orchestration_cache_variable.go:22]: Does `TTLSeconds` accepts `0` as a valid input? Because being so, then `omitempty` will cause a value of `0` for `TTLSeconds` to be omitted during json encoding/decoding.

If this is the case, then using a pointer can workaround this, by intentionally omitting `nil` values travelling in and out through `TTLSeconds`.

File: pagerduty/event_orchestration_cache_variable.go
Comment (imjaroiswebdev) [inline pagerduty/event_orchestration_cache_variable.go]: Please remove the `Context` suffix from methods, it is no longer needed. This apply to the rest of methods too. Thank you!

File: pagerduty/event_orchestration_cache_variable.go
Comment (imjaroiswebdev) [inline pagerduty/event_orchestration_cache_variable.go]: Please prefer to extend `resourceReference` like the rest of resource references in [here](https://github.com/heimweh/go-pagerduty/blob/bfc8dce0a3ffc87473867b75b81c52cb9edd9d9e/pagerduty/references.go#L3).

**Review by imjaroiswebdev — CHANGES_REQUESTED** [review]
@c-kaieong there are a couple of comments to address before moving forward with merge/approval. Other than that, good job 👍🏽

**Comment by imjaroiswebdev** [comment]
> LGTM 👍 Probably worth it to leave it as a draft PR since we can't merge this until the Public API docs go live

I'd also prefer to leave this as a Draft PR in the meantime.

File: pagerduty/event_orchestration_cache_variable.go
Comment (c-kaieong) [reply pagerduty/event_orchestration_cache_variable.go:22]: Hi, `0` is not a valid input so it can be omitted

File: pagerduty/event_orchestration_cache_variable.go
Comment (c-kaieong) [reply pagerduty/event_orchestration_cache_variable.go]: Addressed in https://github.com/heimweh/go-pagerduty/pull/149/commits/cb1dc594997dc4cea3275e5ee1a1f5e2993627be

I'm just directly using the existing User reference instead of creating a new reference.

File: pagerduty/event_orchestration_cache_variable.go
Comment (c-kaieong) [reply pagerduty/event_orchestration_cache_variable.go]: Addressed in https://github.com/heimweh/go-pagerduty/pull/149/commits/cb1dc594997dc4cea3275e5ee1a1f5e2993627be

File: pagerduty/event_orchestration_cache_variable.go
Comment (imjaroiswebdev) [reply pagerduty/event_orchestration_cache_variable.go]: I get why you are doing this, however please add a new type alias for `resourceReference` like the rest of PagerDuty resource are doing.

Comment (c-kaieong) [reply pagerduty/event_orchestration_cache_variable.go]: @imjaroiswebdev Sure I can do that, but just for my own understanding, what's the reasoning for adding a new alias when we'd like to just use a user reference in this case?

Comment (c-kaieong) [reply pagerduty/event_orchestration_cache_variable.go]: I'm assuming I would add the following to [`references.go`](https://github.com/heimweh/go-pagerduty/blob/master/pagerduty/references.go)
```golang
// EventOrchestrationCacheVariableReference represents a reference to an
// Event Orchestration Cache Variable schema
type EventOrchestrationCacheVariableReference resourceReference
```

However, this would not be accurate, since the `CreatedBy` and `UpdatedBy` fields (i.e. the fields using this reference) would be `UserReference`s, no?

Comment (imjaroiswebdev) [reply pagerduty/event_orchestration_cache_variable.go]: Hmmm... now I get what You're saying @c-kaieong, You are right, sorry about the mixup, `CreatedBy` and `UpdatedBy` are indeed `UserReference`s as you are noting, so a new `EventOrchestrationCacheVariableReference` is not needed here.

**Comment by metavida** [comment]
@imjaroiswebdev I just now published pubic REST API docs that are used by this API. For example: https://developer.pagerduty.com/api-reference/a87e8f7117af9-create-a-cache-variable-for-a-global-event-orchestration

The Cache Variables feature in Event Orchestration is now officially GA for all customers so we're ready to merge this PR!

**Review by imjaroiswebdev — APPROVED** [review]
Great! We are ready to merge now 💪🏽

**Pull request merged**
